frontends/common/resolveReferences/resolveReferences.cpp

```diff
@@ -240,186 +216,143 @@ void ResolveReferences::checkShadowing(const IR::INamespace* ns) const {
     if (refMap->checkMap(program))
         return false;
 
     BUG_CHECK(resolveForward.empty(), "Expected empty resolvePath");
     resolveForward.push_back(anyOrder);
-    BUG_CHECK(rootNamespace == nullptr, "Root namespace already set");
-    rootNamespace = program;
-    context = new ResolutionContext(rootNamespace);
     return true;
 }
 
-void ResolveReferences::postorder(const IR::P4Program*) {
-    rootNamespace = nullptr;
-    context->done();
+void ResolveReferences::postorder(const IR::P4Program *) {
     resolveForward.pop_back();
     BUG_CHECK(resolveForward.empty(), "Expected empty resolvePath");
-    context = nullptr;
     LOG1("Reference map " << refMap);
 }
 
-bool ResolveReferences::preorder(const IR::This* pointer) {
+bool ResolveReferences::preorder(const IR::This *pointer) {
     auto decl = findContext<IR::Declaration_Instance>();
     if (findContext<IR::Function>() == nullptr || decl == nullptr)
         ::error("%1%: can only be used in the definition of an abstract method", pointer);
     refMap->setDeclaration(pointer, decl);
     return true;
 }
 
-bool ResolveReferences::preorder(const IR::PathExpression* path) {
-    resolvePath(path->path, false); return true; }
+bool ResolveReferences::preorder(const IR::PathExpression *path) {
+    resolvePath(path->path, false);
+    return true;
+}
 
-bool ResolveReferences::preorder(const IR::Type_Name* type) {
-    resolvePath(type->path, true); return true; }
+bool ResolveReferences::preorder(const IR::Type_Name *type) {
+    resolvePath(type->path, true);
+    return true;
+}
 
 bool ResolveReferences::preorder(const IR::P4Control *c) {
     refMap->usedName(c->name.name);
-    addToContext(c->getTypeParameters());
-    addToContext(c->getApplyParameters());
-    addToContext(c->getConstructorParameters());
-    addToContext(c);  // add the locals
+    checkShadowing(c);
     return true;
 }
 
-void ResolveReferences::postorder(const IR::P4Control *c) {
-    removeFromContext(c);
-    removeFromContext(c->getConstructorParameters());
-    removeFromContext(c->getApplyParameters());
-    removeFromContext(c->getTypeParameters());
-}
-
 bool ResolveReferences::preorder(const IR::P4Parser *p) {
     refMap->usedName(p->name.name);
-    addToContext(p->getTypeParameters());
-    addToContext(p->getApplyParameters());
-    addToContext(p->getConstructorParameters());
-    addToContext(p);
+    checkShadowing(p);
     return true;
 }
 
-void ResolveReferences::postorder(const IR::P4Parser *p) {
-    removeFromContext(p);
-    removeFromContext(p->getConstructorParameters());
-    removeFromContext(p->getApplyParameters());
-    removeFromContext(p->getTypeParameters());
-}
-
-bool ResolveReferences::preorder(const IR::Function* function) {
+bool ResolveReferences::preorder(const IR::Function *function) {
     refMap->usedName(function->name.name);
-    addToContext(function->type->parameters);
+    checkShadowing(function);
```

Comment:
It looks like all these methods now have the same implementation, so perhaps you can factor this in a helper.

Reply:
Yes, I'll look at that.  Unfortunately the visitor infrastructure does not support pre/postorder functions on IR `interface` types, which could make this much cleaner.  I'm thinking about how to extend the ir-generator to fix that.